larq_compute_engine/mlir/python/end2end_verify.cc

```diff
@@ -0,0 +1,118 @@
+#include "tensorflow/lite/optional_debug_tools.h"
+
+// This file is based on the TF lite minimal example
+// Its used in combination with the python/end2end_test.py script
+
+using namespace tflite;
+
+bool getFlatTensor(Interpreter* interpreter, int tensor_id, float** data,
+                   int* num_elements) {
+  const auto* tensor = interpreter->tensor(tensor_id);
+
+  *data = interpreter->typed_tensor<float>(tensor_id);
+
+  if (!*data || tensor->dims == nullptr || tensor->dims->size == 0)
+    return false;
+
+  *num_elements = 1;
+  for (int i = 0; i < tensor->dims->size; ++i)
+    *num_elements *= tensor->dims->data[i];
+
+  return true;
+}
+
+#define TFLITE_MINIMAL_CHECK(x)                             \
+  if (!(x)) {                                               \
+    std::cerr << "Error at line " << __LINE__ << std::endl; \
+    return false;                                           \
+  }
+
+bool verifyModel(const pybind11::bytes& _flatbuffer,
+                 const std::vector<float>& input,
+                 const std::vector<float>& output) {
+  // Load model
+  const auto flatbuffer = static_cast<std::string>(_flatbuffer);
+  std::unique_ptr<tflite::FlatBufferModel> model =
+      tflite::FlatBufferModel::BuildFromBuffer(flatbuffer.data(),
+                                               flatbuffer.size());
+  TFLITE_MINIMAL_CHECK(model != nullptr);
+
+  // Build the interpreter
+  tflite::ops::builtin::BuiltinOpResolver resolver;
+  compute_engine::tflite::RegisterLCECustomOps(&resolver);
+
+  InterpreterBuilder builder(*model, resolver);
+  std::unique_ptr<Interpreter> interpreter;
+  builder(&interpreter);
+  TFLITE_MINIMAL_CHECK(interpreter != nullptr);
+
+  // Allocate tensor buffers.
+  TFLITE_MINIMAL_CHECK(interpreter->AllocateTensors() == kTfLiteOk);
+
+  auto input_ids = interpreter->inputs();
+  auto output_ids = interpreter->outputs();
+
+  TFLITE_MINIMAL_CHECK(input_ids.size() == 1);
+  TFLITE_MINIMAL_CHECK(output_ids.size() == 1);
+
+  float* input_tensor_data;
+  int num_input_elements;
+  TFLITE_MINIMAL_CHECK(getFlatTensor(interpreter.get(), input_ids[0],
+                                     &input_tensor_data, &num_input_elements));
+  TFLITE_MINIMAL_CHECK(num_input_elements == (int)input.size());
+
+  float* output_tensor_data;
+  int num_output_elements;
+  TFLITE_MINIMAL_CHECK(getFlatTensor(interpreter.get(), output_ids[0],
+                                     &output_tensor_data,
+                                     &num_output_elements));
+  TFLITE_MINIMAL_CHECK(num_output_elements == (int)output.size());
+
+  // Do inference twice, to check if the warmpup run and internal temporary
+  // buffers work as expected
+  for (int run = 0; run < 2; ++run) {
+    // Fill input buffers
+    for (int i = 0; i < num_input_elements; ++i) {
+      input_tensor_data[i] = input[i];
+    }
+
+    // Run inference
+    TFLITE_MINIMAL_CHECK(interpreter->Invoke() == kTfLiteOk);
+
+    // Read output buffers
```

Comment:
it would be a lot cleaner if you just return the output buffer to the callee instead of return `true` or `false` bc then you use the python assert/test to check the differences. so instead of having a function `verify_model` you have a function to invoke the inference and get the output back in python callee and you compare using whatever testsuite you are using.

Reply:
Yes. However we want to do two runs on the same interpreter, so then we'd have to concatenate the outputs or something. I'm not sure if thats cleaner, what do you think?

Also I like the output thats generated by the C++ code:
![image](https://user-images.githubusercontent.com/248512/76240587-cb300f00-6233-11ea-8a60-738b89b02c86.png)

Reply:
I don't understand. you are already comparing only the output of the second inference. why do you need to check the both the first and second? For the output format, I would rather stick to the official output format of the testsuite (we can keep this code commented out for debugging purposes though)

Reply:
Currently we compare both outputs.
I'll try to return a list of outputs.

Reply:
It now returns the outputs to python :+1: 
I still want to check both runs, because we want to catch bugs where somehow a warmup run is different from a normal run (maybe a buffer is not ready yet but we're already using it in the first run).